File: server/tests/mood-engine.test.ts
import assert from 'node:assert'
import { describe, it } from 'node:test'

import { clamp, dominantCount, pickMood } from '../src/services/mood-engine.ts'

describe('clamp', () => {
  it('returns value within range', () => {
    assert.strictEqual(clamp(50, 0, 100), 50)
  })

  it('clamps below min', () => {
    assert.strictEqual(clamp(-10, 0, 100), 0)
  })

  it('clamps above max', () => {
    assert.strictEqual(clamp(200, 0, 100), 100)
  })

  it('handles equal bounds', () => {
    assert.strictEqual(clamp(50, 50, 50), 50)
    assert.strictEqual(clamp(0, 50, 50), 50)
  })
})

describe('dominantCount', () => {
  it('counts occurrences', () => {
    assert.strictEqual(dominantCount(['happy', 'sad', 'happy'], 'happy'), 2)
  })

  it('returns 0 for no matches', () => {
    assert.strictEqual(dominantCount(['happy', 'sad'], 'angry'), 0)
  })

  it('works with empty array', () => {
    assert.strictEqual(dominantCount([], 'happy'), 0)
  })
})

describe('pickMood', () => {
  it('severe inactivity -> hungry', () => {
    assert.strictEqual(pickMood(80, [], 0, 24), 'hungry')
  })

  it('moderate inactivity -> tired', () => {
    assert.strictEqual(pickMood(80, [], 0, 12), 'tired')
  })

  it('high score -> laughing', () => {
    assert.strictEqual(pickMood(90, [], 0, 0), 'laughing')
  })

  it('recent chats -> laughing', () => {
    assert.strictEqual(pickMood(70, [], 4, 0), 'laughing')
  })

  it('good score -> happy', () => {
    assert.strictEqual(pickMood(78, [], 0, 0), 'happy')
  })

  it('multiple scared -> scared', () => {
    assert.strictEqual(pickMood(50, ['scared', 'scared'], 0, 0), 'scared')
  })

  it('very low score -> scared', () => {
    assert.strictEqual(pickMood(10, [], 0, 0), 'scared')
  })

  it('multiple angry -> angry', () => {
    assert.strictEqual(pickMood(30, ['angry', 'angry'], 0, 0), 'angry')
  })

  it('moderate score -> angry', () => {
    assert.strictEqual(pickMood(25, [], 0, 0), 'angry')
  })

  it('multiple sad -> sad', () => {
    assert.strictEqual(pickMood(50, ['sad', 'sad'], 0, 0), 'sad')
  })

  it('moderate-low score -> sad', () => {
    assert.strictEqual(pickMood(45, [], 0, 0), 'sad')
  })

  it('defaults to normal', () => {
    assert.strictEqual(pickMood(60, ['happy'], 0, 0), 'normal')
  })

  it('varied moods -> confused', () => {
    assert.strictEqual(pickMood(55, ['happy', 'sad', 'normal'], 0, 0), 'confused')
  })

  it('tired overrides confused when both apply', () => {
    assert.strictEqual(pickMood(55, ['happy', 'sad', 'normal'], 0, 12), 'tired')
  })
})

File: test/mood_engine_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:mochi/db/mochi_repository.dart';
import 'package:mochi/models/mood.dart';

void main() {
  group('clamp', () {
    test('returns value within range', () {
      expect(_clampValue(50, 0, 100), 50);
    });

    test('clamps below min', () {
      expect(_clampValue(-10, 0, 100), 0);
    });

    test('clamps above max', () {
      expect(_clampValue(200, 0, 100), 100);
    });

    test('handles equal bounds', () {
      expect(_clampValue(50, 50, 50), 50);
      expect(_clampValue(0, 50, 50), 50);
    });
  });

  group('dominantCount', () {
    test('counts occurrences', () {
      expect(_dominantCount(<String>['happy', 'sad', 'happy'], 'happy'), 2);
    });

    test('returns 0 for no matches', () {
      expect(_dominantCount(<String>['happy', 'sad'], 'angry'), 0);
    });

    test('works with empty array', () {
      expect(_dominantCount(<String>[], 'happy'), 0);
    });
  });

  group('pickMood', () {
    test('severe inactivity -> hungry', () {
      expect(
        MochiRepository.pickMood(80, <String>[], 0, 24),
        MochiMood.hungry,
      );
    });

    test('moderate inactivity -> tired', () {
      expect(
        MochiRepository.pickMood(80, <String>[], 0, 12),
        MochiMood.tired,
      );
    });

    test('high score -> laughing', () {
      expect(
        MochiRepository.pickMood(90, <String>[], 0, 0),
        MochiMood.laughing,
      );
    });

    test('recent chats -> laughing', () {
      expect(
        MochiRepository.pickMood(70, <String>[], 4, 0),
        MochiMood.laughing,
      );
    });

    test('good score -> happy', () {
      expect(
        MochiRepository.pickMood(78, <String>[], 0, 0),
        MochiMood.happy,
      );
    });

    test('multiple scared -> scared', () {
      expect(
        MochiRepository.pickMood(
          50,
          <String>['scared', 'scared'],
          0,
          0,
        ),
        MochiMood.scared,
      );
    });

    test('very low score -> scared', () {
      expect(
        MochiRepository.pickMood(10, <String>[], 0, 0),
        MochiMood.scared,
      );
    });

    test('multiple angry -> angry', () {
      expect(
        MochiRepository.pickMood(
          30,
          <String>['angry', 'angry'],
          0,
          0,
        ),
        MochiMood.angry,
      );
    });

    test('moderate score -> angry', () {
      expect(
        MochiRepository.pickMood(25, <String>[], 0, 0),
        MochiMood.angry,
      );
    });

    test('multiple sad -> sad', () {
      expect(
        MochiRepository.pickMood(50, <String>['sad', 'sad'], 0, 0),
        MochiMood.sad,
      );
    });

    test('moderate-low score -> sad', () {
      expect(MochiRepository.pickMood(45, <String>[], 0, 0), MochiMood.sad);
    });

    test('defaults to normal', () {
      expect(
        MochiRepository.pickMood(60, <String>['happy'], 0, 0),
        MochiMood.normal,
      );
    });

    test('varied moods -> confused', () {
      expect(
        MochiRepository.pickMood(
          55,
          <String>['happy', 'sad', 'normal'],
          0,
          0,
        ),
        MochiMood.confused,
      );
    });

    test('tired overrides confused when both apply', () {
      expect(
        MochiRepository.pickMood(
          55,
          <String>['happy', 'sad', 'normal'],
          0,
          12,
        ),
        MochiMood.tired,
      );
    });
  });
}

int _clampValue(int value, int min, int max) =>
    value > max ? max : (value < min ? min : value);

int _dominantCount(List<String> moods, String target) =>
    moods.where((String mood) => mood == target).length;
